builtin/plugin/backend.go

```diff
@@ -27,13 +29,76 @@ func Factory(conf *logical.BackendConfig) (logical.Backend, error) {
 // Backend returns an instance of the backend, either as a plugin if external
 // or as a concrete implementation if builtin, casted as logical.Backend.
 func Backend(conf *logical.BackendConfig) (logical.Backend, error) {
+	var b backend
 	name := conf.Config["plugin_name"]
 	sys := conf.System
 
-	b, err := bplugin.NewBackend(name, sys, conf.Logger)
+	raw, err := bplugin.NewBackend(name, sys, conf.Logger)
 	if err != nil {
 		return nil, err
 	}
+	b.Backend = raw
+	b.config = conf
 
-	return b, nil
+	return &b, nil
+}
+
+// backend is a thin wrapper around plugin.BackendPluginClient
+type backend struct {
+	logical.Backend
+	sync.Mutex
+
+	config *logical.BackendConfig
+}
+
+func (b *backend) reloadBackend() error {
+	b.Lock()
+	defer b.Unlock()
+
+	pluginName := b.config.Config["plugin_name"]
+	b.Logger().Trace("plugin: reloading plugin backend", "plugin", pluginName)
+
+	nb, err := bplugin.NewBackend(pluginName, b.config.System, b.config.Logger)
+	if err != nil {
+		return err
+	}
+	err = nb.Setup(b.config)
+	if err != nil {
+		return err
+	}
+	b.Backend = nb
+
+	return nil
+}
+
+// HandleRequest is a thin wrapper implementation of HandleRequest that includes automatic plugin reload.
+func (b *backend) HandleRequest(req *logical.Request) (*logical.Response, error) {
+	resp, err := b.Backend.HandleRequest(req)
```

Comment:
We should grab a read lock prior to calling HandleRequest here, and then give it up before calling reloadBackend. (And grab it again before calling HandleRequest the second time)